### PR TITLE
Test cases

### PR DESCRIPTION
**Pull request by LeandroNz**

### CODE REVIEW

**Review by roxanafalco20 — CHANGES_REQUESTED** [review]
In my opinion, the valid test cases (1 to 6) could be one test case only as the steps/preconditions/posconditions/expected results are the same. The only thing that is changing is the data, but if you take a data driven approach this will be solved. Related to the documentation of the test cases you can add something like “Possible values” and a table with Checking account type | Account ownership | Account name | Initial Deposit Amount.

Same thing with invalids (01 to 10). You can group them by the variable that is changing. I created a table for myself to visualize it better:

￼Screen Shot 2022-03-15 at 16 11 09

Yellow rows are all the same but the Initial Deposit Amount is the only thing different between them. This can be grouped considering the previous approach. Also, Invalid_08 and Invalid_10 are duplicated.

In summary:
- [ ] Consolidate “valid" test cases.
- [ ] Check duplicates on “invalid" test cases.
- [ ] Consolidate “invalid” test cases as much as possible. 
- [ ] Create file/table including all the data for the consolidated test cases

**Review by roxanafalco20 — APPROVED** [review]
Green light on my end :) keep in mind that those possible values should be considered in the automated case.

File: pages/new-checking.page.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    get individualRadioBtn() { return $('#Individual'); }
+    get jointRadioBtn() { return $('#Joint'); }
+    get accountNameInput() { return $('#name'); }
+    get initialDepositInput() { return $('#openingBalance'); }
+    get newCheckingSubmitBtn() { return $('#newCheckingSubmit'); }
+    get newCheckingResetBtn() { return $('#newCheckingReset'); }
+    get formContainer() { return $('.card-body'); }
+    get depositAlertMsg() { return  $('.alert-danger'); }
+    get goHomeBtn() { return $('.home-btn'); }
+    get notFoundErrorContainer() { return $('.notfound-404'); }
```
Comment (axz-abstract) [inline pages/new-checking.page.js:17]: These 2 last selectors are not from this page object, however I see that it is a very particular case to find this 404 page, they can be included in a more general class like the basepage or a common selectors class if they are required again

File: tests/view-checkings.test.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+        let table = ViewCheckingsPage.transactionsTable;
+        await ViewCheckingsPage.navigateTable(4);
+        await ViewCheckingsPage.navigateTable(4);
+
+        await ViewCheckingsPage.clickOnElement(await ViewCheckingsPage.previousBtn);
+        await expect(await browser.checkElement(await table, "wdio-headerContainer", {})).not.toEqual(0);
+
+        await ViewCheckingsPage.clickOnElement(await ViewCheckingsPage.previousBtn);
+        await expect(await browser.checkElement(await table, "wdio-headerContainer", {})).not.toEqual(0);
```
Comment (axz-abstract) [inline tests/view-checkings.test.js:85]: assertions unclear just as in the other test. It should check that the elements are different after every new page loaded, and that the previous or next buttons have the disabled class after reachinf the first or the last page

File: pages/view-checkings.page.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            await expect(this.tableInfoContainer).not.toHaveTextContaining("Showing 1 to 10 of");
+        }
+    }
+
+    /**
+     * Pre: same as super.getNavigationHyperLink function
+     * @param {int} section page or section to navigate to
+     */
+    async navigateTable(section) {
+        await super.clickOnElement(await super.getNavigationHyperLink(section));
```
Comment (axz-abstract) [inline pages/view-checkings.page.js:61]: the getNavigationHyperLink could be in the PO as a dynamic selector

File: pages/base.page.js [hunk 4]
```diff
@@ -2,42 +2,171 @@ const PAGE_TIMEOUT = 10000
...
+     }
+
+     /** Navigate to view checking accounts page
+     *  Pre: user needs to be logged in
+     *  Pos: user is redirected to the create view checking accounts page
+     */
+     async goToViewCheckings() {
+          await this.clickOnElement(this.checkingDropdownMenu);
+          await this.clickOnElement(this.viewCheckingMenuItem);
+     }
```
Comment (axz-abstract) [inline pages/base.page.js:67]: These navigation methods could go in another module dedicated to go to a specific place inside the site, they should be able to do a login as well if possible

File: pages/base.page.js [hunk 5]
```diff
@@ -2,42 +2,171 @@ const PAGE_TIMEOUT = 10000
...
+
+     /**
+      * Pre: index needs to be from an existing hyperlink element
+      * Pos: hyperlink html tag element (a) is returned
+      * @param {int} index number from hyperlink data-dt-idx attribute
+      */
+      async getNavigationHyperLink(index) {
+          return await $('a[data-dt-idx="' + index + '"]');
+     }
```
Comment (axz-abstract) [inline pages/base.page.js:86]: this selector method should go in the view checking PO

File: pages/base.page.js [hunk 6]
```diff
@@ -2,42 +2,171 @@ const PAGE_TIMEOUT = 10000
...
+
+     /** 
+     * Pre: function cannot handle incorrect table selector, 0 <= row <= max row count, 1 <= column <= 5
+     * Pos: text value from td element object is returned
+     * @param {Object} table to get cell from
+     * @param {int} row index to get, index goes from 0 to max row count
+     * @param {int} column index to get, index goes from 1 to 5
+     */
+     async getCellFromTable(table, row, column) {
+          return await table.$$("tr")[row].$("//td[" + column + "]").getText();
```
Comment (axz-abstract) [inline pages/base.page.js:98]: I suggest you change the expression to `$$("tr")[0].$$("th")[1].getText()` as I noticed that you are sending a 0 from the test which would return the table header, therefore making it impossible to find a td inside that header row. when you use the //td you are finding any td in the dom. By using it this way you make sure that you are using the indexes as row and column respectively

File: pages/base.page.js [hunk 7]
```diff
@@ -2,42 +2,171 @@ const PAGE_TIMEOUT = 10000
...
+     async isSortedDescending(arr, sortDate) {
+          let isSorted = true;
+          let arrSorted = null;
+          sortDate ? arrSorted = await arr.sort((a,b) => new Date(b) - new Date(a)) : arrSorted = arr.sort((a,b) => b - a);
+          for (let i = 0; i < arr.length - 1 && isSorted; i++) {
+               isSorted = arr[i] == arrSorted[i];
+          }
+          return isSorted;
+     }
+     //End of Table Helper Functions
```
Comment (axz-abstract) [inline pages/base.page.js:171]: Several methods in this file like fetching an entire column or sorting arrays could be included in a different module of utilities to use with methods with more complexity instead of something more simple such as cleaning an input and entering data for instance

**Review by axz-abstract — CHANGES_REQUESTED** [review]
It needs to do some upgrades but in general terms its pretty good.
Other things I've noticed is that you include login credentials and the login method in each test, this method should be taken to another place to be called in each test such as the wdio.conf where it has other before and after methods to configure behaviors before or after a test.
The credentials could go in a dataprovider file such as the Input JSON files you are using, or you could use them directly in the method I described before, though I prefer the first one and that  they come from a source of data

File: pages/base.page.js [hunk 8]
```diff
@@ -2,42 +2,171 @@ const PAGE_TIMEOUT = 10000
...
+     }
+
+     /** Checks if an array's items are arranged in an descending order
+      * Pre: array length > 1
+      * Pos: returns a boolean value indicating whether the array is correctly sorted or not
+      * @param {Array} arr array to check
+      * @param {boolean} sortDate sort date array or not
+      * @returns boolean
+      */
+     async isSortedDescending(arr, sortDate) {
```
Comment (axz-abstract) [inline pages/base.page.js:162]: I tried these sorting methods and they always returned a true, with an unsorted array and with both true and false as second parameter and they always returned true. Are these working correctly? have you tried them with different sets of data?

File: pages/base.page.js [hunk 9]
```diff
@@ -2,12 +2,13 @@ const PAGE_TIMEOUT = 10000
...
+     get notFoundErrorContainer() { return $('.notfound-404'); }
 
-     //Side menu web elements
-     get checkingDropdownMenu() { return $('#checking-menu'); }
-     get newCheckingMenuItem() { return $('#new-checking-menu-item'); }
-     get viewCheckingMenuItem() { return $('#view-checking-menu-item'); }
-
+     //Alerts
+     get successAlertMsg() { return $('.alert-success'); }
+     get dangerAlertMsg() { return $('.alert-danger'); }
```
Comment (axz-abstract) [inline pages/base.page.js:11]: These are fine for now that there are just a few of them, but if the list of common selectors gets too big you should move them to a separate independent file

File: pages/view-checkings.page.js [hunk 10]
```diff
@@ -34,34 +21,98 @@ class ViewCheckingsPage extends BasePage {
...
+
+    /**
+     * Pre: number needs to be from an existing account
+     * Pos: account card checkbox is click on if it was unchecked
+     * @param {int} account card account number to switch to
+     */
+    async switchAccountCard(account) {
+        let attribute = await $(await this.getElementFromDynamicCardWithNumber(account, 'input')).getAttribute("checked");
+        if (!attribute) {
+            await super.clickOnElement(await this.getElementFromDynamicCardWithNumber(account, 'span'));
```
Comment (axz-abstract) [inline pages/view-checkings.page.js:113]: I like how you use the dynamic methods, however I think `this.getElementFromDynamicCardWithNumber(account, 'input')` and `this.getElementFromDynamicCardWithNumber(account, 'span')` should be included in a method of their own to know exactly that you are finding the account switch and the span to click

File: tests/home.test.js [hunk 11]
```diff
@@ -1,17 +1,8 @@
...
-    //Preconditions
-    beforeEach('Should login', async ()=> {
-        await LoginPage.open('bank');
-        await LoginPage.login(username, password);
-    });
-
-    //Logout test cases
     it('Should logout', async ()=> {
         await HomePage.logout();
         await expect(await LoginPage.successMsg).toBeDisplayed();
```
Comment (axz-abstract) [inline tests/home.test.js:8]: you could include this one with the login tests suite, and use the successMsg in base page rather than LoginPage

File: tests/new-checking.test.js [hunk 12]
```diff
@@ -1,29 +1,26 @@
...
+            let name = await NewCheckingPage.generateUniqueName();
             await NewCheckingPage.createNewCheckingAccount(type, ownership, name, amount);
-            await expect(ViewCheckingsPage.accountCreationConfirmationMsg).toBeDisplayed();
+            await expect(ViewCheckingsPage.successAlertMsg).toBeDisplayed();
+            await expect(ViewCheckingsPage.successAlertMsg).toHaveTextContaining("Successfully created new " + type + " Checking account named " + name);
+            await expect(ViewCheckingsPage.getElementFromDynamicCardWithName(name, ".h4[contenteditable]")).toHaveTextContaining(name);
+            await expect(ViewCheckingsPage.getElementFromDynamicCardWithName(name, "div:nth-child(3) > small")).toHaveTextContaining(type);
+            await expect(ViewCheckingsPage.getElementFromDynamicCardWithName(name, "div:nth-child(4) > small")).toHaveTextContaining(ownership);
+            await expect(ViewCheckingsPage.getElementFromDynamicCardWithName(name, "div:nth-of-type(7)")).toHaveTextContaining(amount);
```
Comment (axz-abstract) [inline tests/new-checking.test.js:23]: Nice to see the `getElementFromDynamicCardWithName` in use, but you should have these selectors in methods inside the page, otherwise you would need to make maintenance of them in this test which is not a good practice, have a descriptive own method for each selector in ViewCheckingsPage

File: tests/new-checking.test.js [hunk 13]
```diff
@@ -34,35 +31,40 @@ describe('New Checking', () => {
...
                     await NewCheckingPage.clickOnElement(NewCheckingPage.goHomeBtn);
+                    await expect(await HomePage.mainMenuNavBar).toBeDisplayed();
                     break;
                 case 24:
-                    await expect(NewCheckingPage.depositAlertMsg).toBeDisplayed();
+                    await expect(NewCheckingPage.dangerAlertMsg).toBeDisplayed();
                     break;
                 default:
-                    await expect(await browser.checkElement(await NewCheckingPage.formContainer, "wdio-headerContainer", {})).not.toEqual(0);
+                    await expect(await ViewCheckingsPage.successAlertMsg).not.toBeDisplayed();
```
Comment (axz-abstract) [inline tests/new-checking.test.js:40]: add a waiting time for this case in which the page will only show an error in the input and not load anything (maybe 5 - 10 seconds at least), moreover you should better use the valid.validity property of the input that hasn't the required pattern, I left something in the Ryver channel on how to do it

File: tests/new-checking.test.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        await NewCheckingPage.sendText(name, 'a');
+        await NewCheckingPage.sendText(amount, 1000);
+
+        let isFormCleared = !(await type.isSelected()) && !(await ownership.isSelected()) && (await name.getValue() == '') && (await amount.getValue() == '');
+        await expect(isFormCleared).toBe(false);
+
+        await NewCheckingPage.clickOnElement(NewCheckingPage.newCheckingResetBtn);
+
+        isFormCleared = !(await type.isSelected()) && !(await ownership.isSelected()) && (await name.getValue() == '') && (await amount.getValue() == '');
+        await expect(isFormCleared).toBe(true);
```
[the remark this answers is not in the document]
Comment (axz-abstract) [reply tests/new-checking.test.js:68]: I see you declared the other radio buttons but you are not clicking them before checking that the form is reset

File: tests/view-checkings.test.js [hunk 15]
```diff
@@ -1,114 +1,96 @@
...
-        let thirdAccount = ViewCheckingsPage.thirdAccountSwitch;
-
-        await ViewCheckingsPage.switchDisplayedAccount(secondAccount);
-        await ViewCheckingsPage.switchDisplayedAccount(firstAccount);
-        await ViewCheckingsPage.switchDisplayedAccount(thirdAccount);
-        await ViewCheckingsPage.switchDisplayedAccount(thirdAccount);
+        let indexs = [2, 1, 3, 3];
+        for (let i = 0; i < indexs.length; i++) {
+            //switch displayed account
+            await ViewCheckingsPage.clickOnElement(await ViewCheckingsPage.getElementFromDynamicCardWithIndex(indexs[i], 'span'));
```
Comment (axz-abstract) [inline tests/view-checkings.test.js:18]: Same case as before for the selector in this method, you should take it to the page object and declare what is you are fetching, you may have a section with different selectors you find inside the card and then reuse them in every `getElementfromDynamicCard...`

File: tests/view-checkings.test.js [hunk 16]
```diff
@@ -1,114 +1,96 @@
...
+            await expect(await browser.checkElement(await ViewCheckingsPage.transactionsTable, "transactionsTable", {})).not.toEqual(0);
```
Comment (axz-abstract) [inline tests/view-checkings.test.js:19]: I would check in here that the table has effectively changed by changing the data in the first row using the Description id of the deposit and the amount, that's a good way to check that the page loaded another table, also I see that the URL  changes with every account selected, you could include that assertion as well

File: tests/view-checkings.test.js [hunk 17]
```diff
@@ -1,114 +1,96 @@
...
+        }
     });
 
     it(`Should not be able to change account name using an empty String`, async () => {
-        await ViewCheckingsPage.changeAccountTitle(ViewCheckingsPage.firstAccountTitleContainer, "");
+        await ViewCheckingsPage.changeAccountTitle(await ViewCheckingsPage.getElementFromDynamicCardWithIndex(1, ".h4[contenteditable]"), "");
```
Comment (axz-abstract) [inline tests/view-checkings.test.js:24]: Same here as my comments on selectors defined in the test

File: tests/view-checkings.test.js [hunk 18]
```diff
@@ -1,114 +1,96 @@
...
         await expect(ViewCheckingsPage.updateNameErrorMsg).toBeDisplayed();
     });
 
     it(`Should be able to change account name using a valid String`, async () => {
-        await ViewCheckingsPage.changeAccountTitle(ViewCheckingsPage.firstAccountTitleContainer, "aaaaaaaaaaaaaaa");
+        await ViewCheckingsPage.changeAccountTitle(await ViewCheckingsPage.getElementFromDynamicCardWithIndex(1, ".h4[contenteditable]"), "aaaaaaaaaaaaaaa");
```
Comment (axz-abstract) [inline tests/view-checkings.test.js:29]: Same here as my comments on selectors defined in the test

File: tests/view-checkings.test.js [hunk 19]
```diff
@@ -1,114 +1,96 @@
...
         await expect(ViewCheckingsPage.updateNameSuccessMsg).toBeDisplayed();
```
Comment (axz-abstract) [inline tests/view-checkings.test.js:30]: Also assert that the same card has effectively changed its name to the new one

File: tests/view-checkings.test.js [hunk 20]
```diff
@@ -1,114 +1,96 @@
...
-        await ViewCheckingsPage.selectTableLength(0);
+        await ViewCheckingsPage.switchAccountCard(SampleAccountNumber.number);
+        let indexs = [1, 0, 2, 0, 3, 0];
+        for (let i = 0; i < indexs.length; i++) {
+            //Select table length
+            await ViewCheckingsPage.tableLengthSelect.selectByIndex(indexs[i]);
+            if (indexs[i] == 0) {
+                await expect(await ViewCheckingsPage.tableInfoContainer).toHaveTextContaining("Showing 1 to 10 of");
+            } else {
+                await expect(await ViewCheckingsPage.tableInfoContainer).not.toHaveTextContaining("Showing 1 to 10 of");
```
Comment (axz-abstract) [inline tests/view-checkings.test.js:42]: Do you have accounts with that many entries? maybe with just one that could have at least 20 and check that when you select the second index it loads the whole entris it would be enough

File: tests/view-checkings.test.js [hunk 21]
```diff
@@ -1,114 +1,96 @@
...
+            let previousColumn = await TableUtils.getColumn(await ViewCheckingsPage.transactionsTable, 0);
+            //Navigate table
+            await ViewCheckingsPage.clickOnElement(await ViewCheckingsPage.getNavigationHyperLink(sections[i]));
+            let actualColumn = await TableUtils.getColumn(await ViewCheckingsPage.transactionsTable, 0);
+            await expect(await ArrayUtils.equalsArray(previousColumn, actualColumn)).toBe(false);
+            //Check previous button on the first section
+            if (await ViewCheckingsPage.isSectionEnabled(1)) await expect(ViewCheckingsPage.previousBtn).toHaveAttributeContaining('class', 'disabled');
+            //Check next button on the last section
+            if (await ViewCheckingsPage.isSectionEnabled(3)) await expect(ViewCheckingsPage.nextBtn).toHaveAttributeContaining('class', 'disabled');
+        }
```
Comment (axz-abstract) [inline tests/view-checkings.test.js:72]: For the pagination numbers its good to use the right index, but for the case of the previous and next buttons have methods to select previous or next buttons

File: tests/view-checkings.test.js [hunk 22]
```diff
@@ -1,114 +1,96 @@
...
-
-        await ViewCheckingsPage.navigateTable(2);
-        await ViewCheckingsPage.navigateTable(3);
-        await ViewCheckingsPage.navigateTable(1);
+        await ViewCheckingsPage.switchAccountCard(SampleAccountNumber.number);
+        let sections = [4, 4, 0, 0, 2, 3, 1];
+        for (let i = 0; i < sections.length; i++) {
+            let previousColumn = await TableUtils.getColumn(await ViewCheckingsPage.transactionsTable, 0);
+            //Navigate table
+            await ViewCheckingsPage.clickOnElement(await ViewCheckingsPage.getNavigationHyperLink(sections[i]));
```
Comment (axz-abstract) [inline tests/view-checkings.test.js:65]: you could wrap this click in a _selectPaginationNumber_ method inside the page object using the index as parameter

File: wdio.conf.js [hunk 23]
```diff
@@ -236,8 +243,12 @@ exports.config = {
...
      * @param {Object} suite suite details
      */
-    // beforeSuite: function (suite) {
-    // },
+    beforeSuite: async function (suite) {
+        if (suite.title != 'Signup' && suite.title != 'Login') {
+            await LoginPage.open('bank');
+            await LoginPage.login(LoginCredentials.username, LoginCredentials.password);
+        }
+    },
```
Comment (axz-abstract) [inline wdio.conf.js:251]: Nice way of defining the Login steps in an upper layer instead of every test, but taking into account the login and sign up cases to not be affected, good job

**Review by axz-abstract — CHANGES_REQUESTED** [review]
2nd review finished

File: pages/view-checkings.page.js [hunk 24]
```diff
@@ -102,16 +140,45 @@ class ViewCheckingsPage extends BasePage {
         return (await $(await this.getNavigationHyperLink(section)).getAttribute("class")).includes("active");
     }
```
Comment (axz-abstract) [inline pages/view-checkings.page.js:141]: this method should be called _isSectionSelected_ or _isSectionActive_ to be more precise

**Review by axz-abstract — CHANGES_REQUESTED** [review]
Nice improvements. 
BTW, are you using a single driver for an entire suite? I ask this because of your `beforeSuite` configuration in _wdio.conf_, if so, then you should change it to run one driver per test, as every test should be independent from each other, except on some special cases

File: pages/view-checkings.page.js [hunk 25]
```diff
@@ -23,7 +52,7 @@ class ViewCheckingsPage extends BasePage {
     async changeAccountTitle(account, title) {
         await super.doubleClickOnElement(account);
         await super.clearEditableContainer(account);
-        await super.sendText(account, title);
+        await browser.keys(title);
```
[the remark this answers is not in the document]
Comment (LeandroNz) [reply pages/view-checkings.page.js:55]: The reason behind it is that when using the setValue() function the container is cleared, but before sending the new text for the title for some reason the container mistakes the empty string as the text for replacement (maybe because setValue() method clicks outside of the container for some reason) and in the end an error message is displayed and the account name is not changed, which is a false error.

File: tests/new-checking.test.js [hunk 26]
```diff
@@ -31,41 +32,29 @@ describe('New Checking', () => {
...
-        && !(await interest.isSelected()) && !(await joint.isSelected()) 
-        && (await name.getValue() == '') && (await amount.getValue() == '');
-        await expect(isFormCleared).toBe(true);
+    RESETINPUT.forEach(({ type, ownership }) => {
+        it(`Should be able to reset the form, selecting the following options: Type -> ${type}, Ownership -> ${ownership}`, async () => {
+            await NewCheckingPage.fillForm(type, ownership, 'a', 1000);
+            await expect(await NewCheckingPage.isFormCleared()).toBe(false);
+            await NewCheckingPage.clickOnElement(NewCheckingPage.newCheckingResetBtn);
+            await expect(await NewCheckingPage.isFormCleared()).toBe(true);
+        });
```
[the remark this answers is not in the document]
Comment (LeandroNz) [reply tests/new-checking.test.js:57]: No, but the create new checking account page is refreshed on each test.

File: pages/base.page.js [hunk 27]
```diff
@@ -68,43 +68,65 @@ export default class BasePage {
...
+      * false if one or more elements are false (vailidity.valid property is false)
       */
      async isFormValid(elements) {
           let isValid = true;
-          for (let i = 0; i < elements.length && isValid; i++) {
-               isValid = await this.checkValidity(await elements[i].getAttribute("id"));
+          for (let group of elements) {
+               let isGroupValid = false;
+               for (let elem of group) isGroupValid = isGroupValid || await this.checkValidity(elem);
+               isValid = isValid && isGroupValid;
```
Comment (axz-abstract) [inline pages/base.page.js:104]: Wouldn't this check that the group is valid if a single element gets the checkValidity on true? just from a simple look I think that isGroupValid attribute would be on true the whole time if it finds a single element which has its validity on true, also why are there 2 loops? are you sending a group of groups of elements?

File: pages/view-checkings.page.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (axz-abstract) [reply pages/view-checkings.page.js:54]: Maybe you could keep trying by sending coordinates of where exactly you want to click the element, debugging could be a nice way to try it out further, but anyways do it just if you wish, if it works this way is fine by me

File: tests/login.test.js [hunk 28]
```diff
@@ -36,8 +33,6 @@ describe('Login', () => {
     });
     
     it('Should be able to login with a valid registered user account and be remembered', async ()=> {      
-        await LoginPage.reload();
-        await LoginPage.open('bank');
         await LoginPage.clickOnElement(LoginPage.rememberMeCheckBox);
         await LoginPage.login(LoginCredentials.username, LoginCredentials.password);
         await expect(await HomePage.mainMenuNavBar).toBeDisplayed();
```
Comment (axz-abstract) [inline tests/login.test.js:38]: In the assertion below I would add a way to check what exactly the cookies are storing and see if the credentials are there, because you don't know exactly what other cookies it may be storing

**Review by axz-abstract — CHANGES_REQUESTED** [review]
4th CR done
